# API下载数据与处理

## 数据来源

[天气API](https://tianqiapi.com/index/doc?version=epidemic)

[实时JSON数据API](http://tianqiapi.com/api?version=epidemic&appid=23035354&appsecret=8YvlPNrz)

<div><img src="./img/0101.png", width=700>

## API请求

In [3]:
import requests
url = 'http://tianqiapi.com/api?version=epidemic&appid=23035354&appsecret=8YvlPNrz'
r = requests.get(url, timeout = 30) # 你要先requests获取整个网页内容
r # 返回200是正常

<Response [200]>

In [4]:
print(r.raise_for_status()) # 如果出错，可以用这个查看错误码

None


In [5]:
r.encoding = r.apparent_encoding # 获取编码格式
r.encoding # 阿斯克码

'ascii'

返回的是ascii，表示他是阿克斯码，需要解码为utf-8

如果用r.json，可以直接解码

In [6]:
import json
data = r.json()
data

{'errcode': 0,
 'data': {'author': '官网:tianqiapi.com 技术QQ:445899710',
  'date': '2020-02-05 18:17:17',
  'diagnosed': 24426,
  'suspect': 23260,
  'death': 493,
  'cured': 1010,
  'list': ['湖北 确诊 16678 例，治愈 540 例，死亡 479 例',
   '广东 确诊 895 例，疑似 139 例，治愈 43 例',
   '浙江 确诊 895 例，治愈 77 例',
   '河南 确诊 764 例，治愈 50 例，死亡 2 例',
   '湖南 确诊 661 例，治愈 50 例',
   '江西 确诊 548 例，治愈 27 例',
   '安徽 确诊 530 例，治愈 23 例',
   '重庆 确诊 376 例，治愈 15 例，死亡 2 例',
   '江苏 确诊 341 例，治愈 23 例',
   '山东 确诊 307 例，治愈 15 例',
   '四川 确诊 301 例，治愈 24 例，死亡 1 例',
   '北京 确诊 253 例，治愈 24 例，死亡 1 例',
   '上海 确诊 243 例，疑似 237 例，治愈 15 例，死亡 1 例',
   '福建 确诊 205 例，治愈 12 例',
   '黑龙江 确诊 190 例，疑似 184 例，治愈 7 例，死亡 2 例',
   '陕西 确诊 165 例，疑似 215 例，治愈 6 例',
   '广西 确诊 150 例，治愈 14 例',
   '河北 确诊 135 例，疑似 117 例，治愈 6 例，死亡 1 例',
   '云南 确诊 124 例，治愈 5 例',
   '海南 确诊 91 例，治愈 5 例，死亡 1 例',
   '辽宁 确诊 81 例，治愈 3 例',
   '山西 确诊 81 例，治愈 5 例',
   '天津 确诊 69 例，治愈 2 例，死亡 1 例',
   '贵州 确诊 64 例，疑似 33 例，治愈 6 例，死亡 1 例',
   '甘肃 确诊 57 例，治愈 4 例',
   '吉林 确诊 54 例，疑似 58 例，治愈 2 例',
   '内蒙古 确诊 4

## 查看数据

既然是字典，查看key值

In [7]:
data.keys()

dict_keys(['errcode', 'data'])

分别查看

In [8]:
data['errcode']

0

明显这个不要

## 数据中的data

In [9]:
Epidemic_data = data['data']
Epidemic_data 

{'author': '官网:tianqiapi.com 技术QQ:445899710',
 'date': '2020-02-05 18:17:17',
 'diagnosed': 24426,
 'suspect': 23260,
 'death': 493,
 'cured': 1010,
 'list': ['湖北 确诊 16678 例，治愈 540 例，死亡 479 例',
  '广东 确诊 895 例，疑似 139 例，治愈 43 例',
  '浙江 确诊 895 例，治愈 77 例',
  '河南 确诊 764 例，治愈 50 例，死亡 2 例',
  '湖南 确诊 661 例，治愈 50 例',
  '江西 确诊 548 例，治愈 27 例',
  '安徽 确诊 530 例，治愈 23 例',
  '重庆 确诊 376 例，治愈 15 例，死亡 2 例',
  '江苏 确诊 341 例，治愈 23 例',
  '山东 确诊 307 例，治愈 15 例',
  '四川 确诊 301 例，治愈 24 例，死亡 1 例',
  '北京 确诊 253 例，治愈 24 例，死亡 1 例',
  '上海 确诊 243 例，疑似 237 例，治愈 15 例，死亡 1 例',
  '福建 确诊 205 例，治愈 12 例',
  '黑龙江 确诊 190 例，疑似 184 例，治愈 7 例，死亡 2 例',
  '陕西 确诊 165 例，疑似 215 例，治愈 6 例',
  '广西 确诊 150 例，治愈 14 例',
  '河北 确诊 135 例，疑似 117 例，治愈 6 例，死亡 1 例',
  '云南 确诊 124 例，治愈 5 例',
  '海南 确诊 91 例，治愈 5 例，死亡 1 例',
  '辽宁 确诊 81 例，治愈 3 例',
  '山西 确诊 81 例，治愈 5 例',
  '天津 确诊 69 例，治愈 2 例，死亡 1 例',
  '贵州 确诊 64 例，疑似 33 例，治愈 6 例，死亡 1 例',
  '甘肃 确诊 57 例，治愈 4 例',
  '吉林 确诊 54 例，疑似 58 例，治愈 2 例',
  '内蒙古 确诊 42 例，治愈 3 例',
  '宁夏 确诊 34 例，治愈 1 例',
  '新疆 确诊 32 例',
  '香

In [10]:
type(Epidemic_data)

dict

查看里面的key值

In [11]:
Epidemic_data.keys()

dict_keys(['author', 'date', 'diagnosed', 'suspect', 'death', 'cured', 'list', 'history', 'area'])

观察到其中：

'date', 'diagnosed', 'suspect', 'death', 'cured', 'list' 比较简单，为当日汇总总数

'history'和'area'比较复杂

### 全国历史数据及折线图

查看'history'

In [12]:
history = Epidemic_data['history']
history

[{'date': '2020-02-04',
  'confirmedNum': 24363,
  'suspectedNum': 23260,
  'curesNum': 892,
  'deathsNum': 491,
  'suspectedIncr': 3971},
 {'date': '2020-02-03',
  'confirmedNum': 20471,
  'suspectedNum': 23214,
  'curesNum': 630,
  'deathsNum': 425,
  'suspectedIncr': 5072},
 {'date': '2020-02-02',
  'confirmedNum': 17238,
  'suspectedNum': 21558,
  'curesNum': 475,
  'deathsNum': 361,
  'suspectedIncr': 5173},
 {'date': '2020-02-01',
  'confirmedNum': 14411,
  'suspectedNum': 19544,
  'curesNum': 328,
  'deathsNum': 304,
  'suspectedIncr': 4562},
 {'date': '2020-01-31',
  'confirmedNum': 11821,
  'suspectedNum': 17988,
  'curesNum': 243,
  'deathsNum': 259,
  'suspectedIncr': 5019},
 {'date': '2020-01-30',
  'confirmedNum': 9720,
  'suspectedNum': 15238,
  'curesNum': 171,
  'deathsNum': 213,
  'suspectedIncr': 4812},
 {'date': '2020-01-29',
  'confirmedNum': 7736,
  'suspectedNum': 12167,
  'curesNum': 124,
  'deathsNum': 170,
  'suspectedIncr': 4148},
 {'date': '2020-01-28',
  'co

这个应该是每日全国总数，可以直接画出折线图，其中数据为

“证实数”: 14482,

“治愈数”: 344,

“日期”:“2020-02-01”,

“死亡数”: 304,

“疑似数”: 19544


In [13]:
len(history)

13

此处应有plotly折线图

In [14]:
dict_history = {'confirmedNum': [],
       'curesNum': [],
       'date': [],
       'deathsNum': [],
       'suspectedNum': [],
        'suspectedIncr': []}
for data_day in history:
    for key, values in data_day.items():
        dict_history[key].append(values)
dict_history

{'confirmedNum': [24363,
  20471,
  17238,
  14411,
  11821,
  9720,
  7736,
  5997,
  4535,
  2761,
  1985,
  1297,
  830],
 'curesNum': [892, 630, 475, 328, 243, 171, 124, 103, 51, 49, 38, 38, 34],
 'date': ['2020-02-04',
  '2020-02-03',
  '2020-02-02',
  '2020-02-01',
  '2020-01-31',
  '2020-01-30',
  '2020-01-29',
  '2020-01-28',
  '2020-01-27',
  '2020-01-26',
  '2020-01-25',
  '2020-01-24',
  '2020-01-23'],
 'deathsNum': [491, 425, 361, 304, 259, 213, 170, 132, 106, 80, 56, 41, 25],
 'suspectedNum': [23260,
  23214,
  21558,
  19544,
  17988,
  15238,
  12167,
  9239,
  6973,
  5794,
  2684,
  1965,
  1072],
 'suspectedIncr': [3971,
  5072,
  5173,
  4562,
  5019,
  4812,
  4148,
  3248,
  2077,
  3806,
  1309,
  1118,
  680]}

In [15]:
df_history = pd.DataFrame(dict_history)
df_history.set_index(['date'], inplace=True)

In [17]:
import matplotlib as mpl
import matplotlib.pyplot as plt  
%matplotlib inline
import numpy as np
import pylab
import pandas as pd
import plotly.offline as plyo  
plyo.iplot(  
    df_history.iplot(asFigure=True),  
    image='png',  
    filename='ply_01'  
)

## 地区累计详细数据

In [18]:
Country_Data = data['data']['area']
Country_Data

[{'provinceName': '湖北',
  'provinceShortName': '湖北',
  'confirmedCount': 16678,
  'suspectedCount': 0,
  'curedCount': 540,
  'deadCount': 479,
  'comment': '',
  'locationId': 420000,
  'cities': [{'cityName': '武汉',
    'confirmedCount': 8351,
    'suspectedCount': 0,
    'curedCount': 373,
    'deadCount': 362},
   {'cityName': '黄冈',
    'confirmedCount': 1645,
    'suspectedCount': 0,
    'curedCount': 52,
    'deadCount': 25},
   {'cityName': '孝感',
    'confirmedCount': 1462,
    'suspectedCount': 0,
    'curedCount': 9,
    'deadCount': 18},
   {'cityName': '襄阳',
    'confirmedCount': 735,
    'suspectedCount': 0,
    'curedCount': 7,
    'deadCount': 2},
   {'cityName': '荆州',
    'confirmedCount': 713,
    'suspectedCount': 0,
    'curedCount': 12,
    'deadCount': 9},
   {'cityName': '随州',
    'confirmedCount': 706,
    'suspectedCount': 0,
    'curedCount': 9,
    'deadCount': 8},
   {'cityName': '黄石',
    'confirmedCount': 509,
    'suspectedCount': 0,
    'curedCount': 23,
  

我们的目标结构

![img_0201](.\img\0102.png)


BOB写的代码

In [42]:
def city(province):
    '''
    处理港澳地区无城市的情况
    province: 类型:字典型
    return:   包含以下内容的 list
    '''
    d = {}                                          #创建字典保存数据
    #d['province'] = province['provinceName']
    try:
        d['cityName'] = province['provinceName']        #获取省份名称
    except:
        d['cityName'] = province['cityName'] # 石河子只有cityName没有provinceCity
    d['confirmed'] = province['confirmedCount']     #确诊数
    d['suspected'] = province['suspectedCount']     #疑似数
    d['cured'] = province['curedCount']             #治愈数
    d['dead'] = province['deadCount']               #死亡数
    
    return list(d.values())                         #返回上述内容list

In [43]:
def requestData():
    '''
    获取最新疫情数据, 保存至当前目录为 日期.csv 文件
    return: 文件名(日期.csv)
    '''
    #数据更新网址
    url = 'http://tianqiapi.com/api?version=epidemic&appid=23035354&appsecret=8YvlPNrz'
    #返回 json 格式数据
    data = pd.read_json(url, orient='colums')
    
    #提取到总数据 待提取相关详细数据
    Epidemic_data = data['data'] 
    #数据日期
    Data_date = Epidemic_data['date'].split()[0]
    #columns = ['provinceName', 'cityName', 'confirmed', 'suspected', 'cured', 'dead']
    columns = ['cityName', 'confirmed', 'suspected', 'cured', 'dead']
    
    #创建全国疫情数据 DataFrame
    China_Data = pd.DataFrame(columns=['provinceName', 
                                       'cityName', 
                                       'confirmed', 
                                       'suspected', 
                                       'cured', 
                                       'dead'])
    #获取所有省份数据
    Country_Data = Epidemic_data['area']
    
    for province in Country_Data:                           #遍历每一个省份
        provinceName = province['preProvinceName']          #获取省份名称
        try:
            ProvinceData = province['cities']               #获取该省份下所有城市数据
            city_Data = []
            for province in ProvinceData:                   #遍历所有城市
                city_Data.append(list(province.values()))
            #构建 DataFrame 格式
            CityData = pd.DataFrame(np.array(city_Data), columns=columns)
            #新增省份列属性, 设置省份名称
            CityData['provinceName'] = provinceName
            #更新至总数据中
            China_Data = pd.concat([China_Data, CityData], ignore_index=True)
        except:#处理香港, 澳门等单一城市的数据 代码类似
            CityData = pd.DataFrame(np.array([city(province)]), columns=columns)
            CityData['provinceName'] = provinceName
            China_Data = pd.concat([China_Data, CityData], ignore_index=True)
            #print(CityData)
            #print(pd.DataFrame(np.array([city(province)]), columns=columns))
            #none_City.append(city(province))
    
    #将省份和城市作分级 保存 csv
    China_Data.set_index(['provinceName', 'cityName'], inplace=True)
    
    #保存文件名称以数据日期为准
    fileName = Data_date + '.csv'
    China_Data.to_csv(fileName)
    return fileName

In [44]:
fileName = requestData()

D:\anaconda\lib\site-packages\ipykernel_launcher.py:40: FutureWarning:

Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.


D:\anaconda\lib\site-packages\ipykernel_launcher.py:44: FutureWarning:

Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.




In [45]:
fileName

'2020-02-05.csv'

使用Epidemic_Data.py整理

In [46]:
data = pd.read_csv(fileName)

In [47]:
data.set_index(['provinceName', 'cityName'], inplace=True)

In [27]:
data.head(30)

confirmed  cured  dead  suspected
provinceName cityName                                   
湖北省          武汉             8351    373   362          0
             黄冈             1645     52    25          0
             孝感             1462      9    18          0
             襄阳              735      7     2          0
             荆州              713     12     9          0
             随州              706      9     8          0
             黄石              509     23     2          0
             宜昌              496      9     4          0
             荆门              422     16    16          0
             咸宁              384      3     0          0
             鄂州              382      6    18          0
             十堰              318      9     0          0
             仙桃              225      0     4          0
             恩施州             138     10     0          0
             天门              128      0    10          0
             潜江               54      0     1          0
             神农架林区            10      2     0          0
广东省          深圳              291     16     0          0
             广州              249      7     0          0
             珠海               72      4     0          0
             佛山               50      1     0          0
             东莞               44      1     0          0
             中山               36      3     0          0
             惠州               34      1     0          0
             汕头               20      3     0          0
             湛江               18      3     0          0
             肇庆               13      1     0          0
             阳江               12      0     0          0
             梅州               10      0     0          0
             清远               10      2     0          0

## 如何排查报错

在jupyter中手动

In [28]:
def city(province):
    '''
    处理港澳地区无城市的情况
    province: 类型:字典型
    return:   包含以下内容的 list
    '''
    d = {}                                          #创建字典保存数据
    d['province'] = province['provinceName']
    d['cityName'] = province['provinceName']        #获取省份名称
    d['confirmed'] = province['confirmedCount']     #确诊数
    d['suspected'] = province['suspectedCount']     #疑似数
    d['cured'] = province['curedCount']             #治愈数
    d['dead'] = province['deadCount']               #死亡数
    
    return list(d.values())                         #返回上述内容list

In [30]:
def city(province):
    '''
    处理港澳地区无城市的情况
    province: 类型:字典型
    return:   包含以下内容的 list
    '''
    d = {}                                          #创建字典保存数据
    #d['province'] = province['provinceName']
    try:
        d['cityName'] = province['provinceName']        #获取省份名称
    except:
        print(province)
    d['confirmed'] = province['confirmedCount']     #确诊数
    d['suspected'] = province['suspectedCount']     #疑似数
    d['cured'] = province['curedCount']             #治愈数
    d['dead'] = province['deadCount']               #死亡数
    
    return list(d.values())                         #返回上述内容list

石河子只有cityName没有provinceCity

注：如果使用spyder

spyder 里面 右上角 有个变量器

<div><img src="./img/0103.png", width=900>
    
以此排开... 然后看变量

<div><img src="./img/0104.png", width=900>

## 地理坐标

可以使用API获得经纬度数据

[使用Python中的folium包创建热力密度图 
原创 杜雨 数据小魔方 2018-06-07](https://mp.weixin.qq.com/s?__biz=MzA3Njc0NzA0MA==&mid=2653192478&idx=1&sn=b5ccc5101c39310c57b9cd45dd7996f9&chksm=848cb951b3fb30472a66ac7122319fb0e0280591cbb37718372cdf1768136f802af5989e9c97&scene=21#wechat_redirect)

In [34]:
import folium
import  time
import requests
from urllib.request import quote
import numpy as np
import pandas as pd
import seaborn as sns
import webbrowser
from folium.plugins import HeatMap

In [35]:
url = "http://api.map.baidu.com/geocoder/v2/"
header = {'User-Agent':'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.79 Safari/537.36'}
payload = {
    'output':'json',
    'ak':'X8zlxPUdSe2weshrZ1WqnWxb43cfBI2N'
    }
addinfo = []
payload['address'] = '北京'
content =  requests.get(url,params=payload,headers=header).json()

In [36]:
content

{'status': 0,
 'result': {'location': {'lng': 116.4133836971231, 'lat': 39.910924547299565},
  'precise': 0,
  'confidence': 20,
  'comprehension': 100,
  'level': '城市'}}

In [37]:
content['result']['location']

{'lng': 116.4133836971231, 'lat': 39.910924547299565}

### 批量匹配经纬度数据

In [38]:
def getCityLocation(cityList):
    '''
    用于获取城市经纬度
    cityList: 全国城市名称的 list
    return: 城市经纬度的 DataFrame
    '''
    url = "http://api.map.baidu.com/geocoder/v2/"
    header = {'User-Agent':'Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.79 Safari/537.36'}
    payload = {
        'output':'json',
        'ak':'X8zlxPUdSe2weshrZ1WqnWxb43cfBI2N'
        }
    addinfo = []
    for city in cityList:
        payload['address'] = city
        try:
            content =  requests.get(url,params=payload,headers=header).json()
            longitude = content['result']['location']['lng']
            latitude = content['result']['location']['lat']
            addinfo.append([city, longitude, latitude])
            #addinfo.append(content['result']['location'])
            print("正在获取{}的地址！".format(city))        
        except:
            print("地址{}获取失败，请稍后重试！".format(city))    
            pass
        #time.sleep(.1)
        
    columns = ['cityName', 'longitude', 'latitude']
    data = pd.DataFrame(np.array(addinfo), columns=columns)
    #print(data)
    print("所有地址均已获取完毕！！！")
    return(data)

In [48]:
data = pd.read_csv(fileName)
data.head()

provinceName cityName  confirmed  cured  dead  suspected
0          湖北省       武汉       8351    373   362          0
1          湖北省       黄冈       1645     52    25          0
2          湖北省       孝感       1462      9    18          0
3          湖北省       襄阳        735      7     2          0
4          湖北省       荆州        713     12     9          0

In [49]:
cityList = data['cityName'].tolist()    #获取所有城市列表

In [50]:
location = getCityLocation(cityList)    #获取城市经纬度
UpdateData = pd.merge(data, location,   #更新经纬度
                      on='cityName', 
                      how = 'left')

正在获取武汉的地址！
正在获取黄冈的地址！
正在获取孝感的地址！
正在获取襄阳的地址！
正在获取荆州的地址！
正在获取随州的地址！
正在获取黄石的地址！
正在获取宜昌的地址！
正在获取荆门的地址！
正在获取咸宁的地址！
正在获取鄂州的地址！
正在获取十堰的地址！
正在获取仙桃的地址！
正在获取恩施州的地址！
正在获取天门的地址！
正在获取潜江的地址！
正在获取神农架林区的地址！
正在获取深圳的地址！
正在获取广州的地址！
正在获取珠海的地址！
正在获取佛山的地址！
正在获取东莞的地址！
正在获取中山的地址！
正在获取惠州的地址！
正在获取汕头的地址！
正在获取湛江的地址！
正在获取肇庆的地址！
正在获取阳江的地址！
正在获取梅州的地址！
正在获取清远的地址！
正在获取江门的地址！
正在获取韶关的地址！
正在获取茂名的地址！
正在获取揭阳的地址！
正在获取汕尾的地址！
正在获取潮州的地址！
正在获取河源的地址！
正在获取温州的地址！
正在获取杭州的地址！
正在获取台州的地址！
正在获取宁波的地址！
正在获取金华的地址！
正在获取绍兴的地址！
正在获取嘉兴的地址！
正在获取丽水的地址！
正在获取衢州的地址！
正在获取湖州的地址！
正在获取舟山的地址！
正在获取信阳的地址！
正在获取南阳的地址！
正在获取郑州的地址！
正在获取驻马店的地址！
正在获取商丘的地址！
正在获取周口的地址！
正在获取新乡的地址！
正在获取安阳市的地址！
正在获取平顶山的地址！
正在获取许昌的地址！
正在获取漯河市的地址！
正在获取开封的地址！
正在获取洛阳的地址！
正在获取焦作的地址！
正在获取鹤壁的地址！
正在获取三门峡的地址！
正在获取濮阳的地址！
正在获取济源的地址！
正在获取长沙的地址！
正在获取岳阳的地址！
正在获取邵阳的地址！
正在获取常德的地址！
正在获取娄底的地址！
正在获取益阳的地址！
正在获取衡阳的地址！
正在获取株洲的地址！
正在获取怀化的地址！
正在获取永州的地址！
正在获取郴州的地址！
正在获取湘潭的地址！
地址湘西自治州获取失败，请稍后重试！
正在获取张家界的地址！
正在获取南昌的地址！
正在获取九江的地址！
正在获取新余的地址！
正在获取上饶的地址！
正在获取赣州的地址！
正在获取宜春的地址！
正在获取抚州的地址！
正在获取萍乡的地址！
正在获取吉安的地址！
正在获

In [51]:
UpdateData.head()

provinceName cityName  confirmed  cured  dead  suspected  \
0          湖北省       武汉       8351    373   362          0   
1          湖北省       黄冈       1645     52    25          0   
2          湖北省       孝感       1462      9    18          0   
3          湖北省       襄阳        735      7     2          0   
4          湖北省       荆州        713     12     9          0   

            longitude            latitude  
0  114.31158155473231  30.598466736400987  
1  114.87849048410779    30.4593588576181  
2  113.92251007733665  30.930689227018295  
3  112.12853720100244  32.014796804669224  
4  112.24552262926137  30.340842107742912

保存数据

In [52]:
UpdateData.to_csv('Visualization_Data.csv')

其他数据源推荐

[无码科技对外开放发热门诊和定点医院数据接口](https://mp.weixin.qq.com/s/NazhWobs6EQCageEyO5LCQ)

[Kaggle数据集]()